### ISSUE TITLE
Allow IMEX solvers for stiff problems

### ISSUE THREAD

**Issue by JoshuaLampert** [JoshuaLampert/DispersiveShallowWater.jl#165]
Currently, we wrap every semidiscretization in an `ODEProblem`. However, some equations contain stiff terms, such as the Svärd-Kalisch equations with $\alpha = 0$ and/or $\gamma = 0$ or the KdV equation (once implemented). Therefore, it would be nice to split up the rhs into a non-stiff part and a stiff part, which are then stored as a `SplitODEProblem`. This allows using IMEX time integrators.

**Comment by JoshuaLampert** [comment]
Two additional thoughts:
- Do we want to create a `SplitODEProblem` for every equation, i.e. also for those without stiff part by setting the stiff function simply to zero or do we want to stick an `ODEProblem` for these equations? In the latter case we could probably add a trait `has_stiff_terms(equations)` or something like that and dispatch on that. Do you see a disadvantage of the first option, @ranocha?
- If we want to use a `SplitODEProblem` for the Svärd-Kalisch equations, we probably need to use the `mass_matrix` argument from the `SplitFunction` instead of explicitly solving the linear system by our own. Is there a problem with that or is this maybe even benefitial?

**Comment by ranocha** [comment]
Setting up a `SplitODEProblem` for non-stiff problems has an additional overhead. I don't know how much but we should investigate this before deciding how to proceed. Conceptually, it would be simpler if an ODE problem can just be an `ODEProblem`.

**Comment by ranocha** [comment]
I have not used the mass matrix formulation lately, so we need to check. However, we need to ensure that it works as desired, since the mass matrix depends on the state and does not occur (in a non-trivial form) for all components.

**Comment by JoshuaLampert** [comment]
> I have not used the mass matrix formulation lately, so we need to check. However, we need to ensure that it works as desired, since the mass matrix depends on the state and does not occur (in a non-trivial form) for all components.

Yes, that's true. However, it might work with a `MatrixOperator` from SciMLOperators.jl, see https://docs.sciml.ai/SciMLOperators/stable/interface/#Operator-update. We can provide an `update_func` taking the state variables. If and how well that works would need to be tested of course. See also https://discourse.julialang.org/t/state-dependent-mass-matrix/40238/5.
